roachpb/batch.go

```diff
@@ -176,11 +176,264 @@ func (ba *BatchRequest) Methods() []Method {
 }
 
 // CreateReply creates replies for each of the contained requests, wrapped in a
-// BatchResponse.
+// BatchResponse. The response objects are batch allocated to minimize
+// allocation overhead which adds a bit of complexity to this function.
 func (ba *BatchRequest) CreateReply() *BatchResponse {
 	br := &BatchResponse{}
+	br.Responses = make([]ResponseUnion, len(ba.Requests))
+
+	var counts struct {
```

Comment:
You could add a simple test which reflects over the fields in `RequestUnion` and promptly alerts users that something's missing here (i.e. for each pointer field in `RequestUnion`, create a batch with only one of those requests, call `CreateReply`, fail unless there's exactly one request in there).